### PR TITLE
HTTP/3: Fix accepting simultaneous connections

### LINKED ISSUE #73282: HTTP/3: Error when multiple connections to a server
### Description

I've noticed the gRPC benchmark has failed requests when there are multiple connections.

### Reproduction Steps

1. `git clone https://github.com/jamesnk/grpc-dotnet`
2. `git checkout jamesnk/net7`
3. `dotnet run -c Release --project .\perf\benchmarkapps\GrpcAspNetCoreServer\ -- --LogLevel Trace --protocol h3`
4. `dotnet run -c Release --project .\perf\benchmarkapps\GrpcClient\ -- -u https://localhost:5000 -c 2 --streams 1 -s unary -p h3 --grpcClientType grpcnetclient --LogLevel Trace --callCount 10`

### Expected behavior

Requests all succeed.

### Actual behavior

One or two requests fail. Also, the client app takes a long time (20 seconds?) to close. I think a request on one of the connections is hung.

Client and server logs are attached.

* [server.log](https://github.com/dotnet/runtime/files/9248643/server.log)
* [client.log](https://github.com/dotnet/runtime/files/9248644/client.log)

### Regression?

I think so.

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/runtime/issues/73282

The previous way of capturing context doesn't work because the connection callback is executed in a background thread and not when `AcceptConnectionAsync` is called.

So, what do we think of using a weak table here?

### CODE REVIEW

**Comment by halter73** [comment]
> So, what do we think of using a weak table here?

I don't love it. Do we have a request for the msquic and/or runtime library to somehow notify us when it calls our `ConnectionOptionsCallback` but then fails to continue accepting the connection for some reason? It shouldn't be necessary to resort to weak references for this.

**Comment by halter73** [comment]
I remember. The idea was to have the `QuicConnection` hold onto the reference for us. Where's the issue for that?

**Comment by JamesNK** [comment]
QuicConnection property bag https://github.com/dotnet/runtime/issues/72511

File: src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs [hunk 1]
```diff
@@ -134,8 +140,14 @@ public async ValueTask CreateListenerAsync()
         {
             var quicConnection = await _listener.AcceptConnectionAsync(cancellationToken);
 
-            // _currentAcceptingConnection is set inside ConnectionOptionsCallback.
-            var connectionContext = _currentAcceptingConnection;
+            if (!_pendingConnections.TryGetValue(quicConnection, out var connectionContext))
+            {
+                throw new InvalidOperationException("Couldn't find ConnectionContext for QuicConnection.");
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:145]: What's keeping the `QuicConnectionContext` rooted until we get to this point? It has no more strong references as soon as `await _tlsConnectionCallbackOptions.OnConnection(context, cancellationToken);` completes, right?

**Comment by halter73** [comment]
> QuicConnection property bag [dotnet/runtime#72511](https://github.com/dotnet/runtime/issues/72511)

Can we mark that `ready-for-review` and `blocking`? 😄

**Comment by JamesNK** [comment]
> > So, what do we think of using a weak table here?
> 
> I don't love it. Do we have a request for the msquic and/or runtime library to somehow notify us when it calls our `ConnectionOptionsCallback` but then fails to continue accepting the connection for some reason? It shouldn't be necessary to resort to weak references for this.

I believe `AcceptConnectionAsync` always throws an error, but we don't know which `QuicConnection` caused the error so can't remove it from the table.

File: src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:145]: The CWT is keeping `QuicConnectionContext` rooted. The important thing to keep rooted so the CWT isn't emptied is `QuicConnection`.

A couple of lines above this:

```
var quicConnection = await _listener.AcceptConnectionAsync(cancellationToken);
```

Comment (halter73) [reply src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:145]: Right. I forgot how CWT works lol.

**Comment by JamesNK** [comment]
> > QuicConnection property bag [dotnet/runtime#72511](https://github.com/dotnet/runtime/issues/72511)
> 
> Can we mark that `ready-for-review` and `blocking`? 😄

That's not our call to make. Need to convince networking to take it first. Do you want to avoid this approach and push adding a property bag in .NET 7?

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
> That's not our call to make. Need to convince networking to take it first. Do you want to avoid this approach and push adding a property bag in .NET 7?

Maybe it's not so bad. It can probably wait.

**Pull request merged**